tests/disabled_user_test.py

```diff
@@ -0,0 +1,39 @@
+# -*- coding: utf-8 -*-
+"""Test auth_role.DISABLE and api_perm.require_not_disabled
+
+:copyright: Copyright (c) 2024 RadiaSoft LLC.  All Rights Reserved.
+:license: http://www.apache.org/licenses/LICENSE-2.0.html
+"""
+import pytest
+
+
+def test_disabled_user(fc):
+    from pykern.pkcollections import PKDict
+    from pykern.pkunit import pkeq, pkok
+    import sirepo.auth_role
+    import sirepo.pkcli.roles
+
+    a = fc.sr_auth_state()
```

Comment:
This is actually interesting. We don't necessarily want to return this particular role in auth_state, because it gives an attacker a reason for the forbidden, a way to probe the system. Therefore, auth_state will have to filter this out.

Reply:
With our code being open source and this PR being public it seems like they could figure it out regardless

Reply:
I understand that, but someone could still look through our code/conversations and suspect that they have this role. I don't see how filtering out this role from auth_state makes things any more secure.

Reply:
Here's a new thought: when two programmers disagree, there's something wrong with the design. @rorour, you discovered a problem. The original solution to use "negative role" is an incorrect hack. What we would do in a password-based systems is disable the password so they couldn't login any more. Even that does not give a reason or have any audit support. I chose that solution because we have cli's that support this already. Maybe when two programmers disagree, the customer is being lazy (or cheap :). The right solution is more expensive, or is it?

With moderation, we deny and have a person bound to it with a time. This is a similar problem, but it's different. If we were doing true RBAC, we would have a permission to use the system at all. We could have a db upgrade that adds a role "user" or something like that. require_user would check this role, because that's the surrogate. (In BOP, we had such a role as all access, even [visitors got a role and permissions](https://github.com/biviosoftware/perl-Bivio/blob/master/Util/SQL.pm#L962).) 

We could also add then a UserRoleInvite record when we deny the user with a simple command line utility. The login in require_user would be the same, but it would be positive: in order to validate as a user there would have to be a role. auth.login wouldn't allow the user to login either. Perhaps that's giving information that they are denied, but as you note, they will be able to detect when they are denied in any event, and infer something. The key is that we don't return authState with roles unless they are logged in so they don't really get any more information with auth_state.

Thoughts?

Reply:
Yes, I think making the role positive rather than negative is a great improvement on this idea.

Would new users have to wait for approval?

Reply:
> Would new users have to wait for approval?

We won't build this yet, but this might be an option. For now, all users will get the role on creation.